crates/runtime/src/dataconnector/github/github_app_token_wrapper.rs

```diff
@@ -0,0 +1,172 @@
+    UnableToGetGitHubInstallationAccessToken { source: reqwest::Error },
+
+    #[snafu(display("Unable to get GitHub installation access token body"))]
+    UnableToGetGitHubInstallationAccessTokenBody { source: reqwest::Error },
+}
+
+#[derive(Serialize)]
+struct Claims {
+    iat: usize,
+    exp: usize,
+    iss: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct InstallationTokenResponse {
+    token: String,
+    expires_at: String,
+}
+
+pub struct GitHubAppTokenWrapper {
+    token: Arc<RwLock<String>>,
+    expires_at: Arc<RwLock<String>>,
+    app_client_id: Arc<str>,
+    private_key: Arc<str>,
+    installation_id: Arc<str>,
+}
+
+impl GitHubAppTokenWrapper {
+    #[must_use]
+    pub fn new(app_client_id: Arc<str>, private_key: Arc<str>, installation_id: Arc<str>) -> Self {
+        Self {
+            token: Arc::new(RwLock::new(String::new())),
+            expires_at: Arc::new(RwLock::new(String::new())),
+            app_client_id,
+            private_key,
+            installation_id,
+        }
+    }
+
+    async fn generate_token(&self) -> Result<String, GitHubAppError> {
+        let iat = usize::try_from(
+            SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .context(UnableToGetSystemTimeSnafu {})?
+                .as_secs(),
+        )
+        .context(InvalidSystemTimeSnafu {})?;
+
+        let exp = iat + 600;
+        let claims = Claims {
+            iat,
+            exp,
+            iss: self.app_client_id.to_string(),
+        };
+        let private_key = self.private_key.as_ref();
+        let encoding_key =
+            EncodingKey::from_rsa_pem(private_key.as_bytes()).context(InvalidPrivateKeySnafu {})?;
+
+        let jwt_token = encode(&Header::new(Algorithm::RS256), &claims, &encoding_key)
+            .context(UnableToGenerateJWTSnafu {})?;
+
+        let client = reqwest::Client::new();
+
+        let response = client
+            .post(format!(
+                "https://api.github.com/app/installations/{}/access_tokens",
+                self.installation_id
+            ))
+            .header("Accept", "application/vnd.github+json")
+            .header("Authorization", format!("Bearer {jwt_token}"))
+            .header("X-GitHub-Api-Version", "2022-11-28")
+            .header("User-Agent", "spice")
+            .send()
+            .await
+            .context(UnableToGetGitHubInstallationAccessTokenSnafu {})?;
+
+        let token_response: InstallationTokenResponse = response
+            .json()
+            .await
+            .context(UnableToGetGitHubInstallationAccessTokenBodySnafu {})?;
+
+        self.expires_at
+            .write()
+            .await
+            .clone_from(&token_response.expires_at);
+
+        Ok(token_response.token)
+    }
+}
+
+#[async_trait]
+impl TokenWrapper for GitHubAppTokenWrapper {
+    fn is_refreshable(&self) -> bool {
+        true
+    }
+
+    async fn get_token(&self) -> Result<String> {
```

Comment:
Todo: add tests. Need to make sure that it handles concurrent requests with nearly expiring token correctly